test/ts/index.tsx

```diff
@@ -1,41 +1,94 @@
-import { h, app } from "hyperapp"
+import { h, app } from "hyperapp";
+
+// Module 1
+
+interface Module1State {
+  count: number;
+}
+
+interface Module1Actions extends Hyperapp.Actions {
+  sub(value: number): Partial<Module1State>;
+  add(value: number): Partial<Module1State>;
+}
+
+const module1Actions: Hyperapp.ActionsImpl<Module1State, Module1Actions> = {
+  sub: (state, actions, value: number) => ({ count: state.count - value }),
+  add: (state, actions, value: number) => ({ count: state.count + value })
+};
+
+// Module 2
+
+interface Module2State {
+  count: number;
+}
+
+interface Module2Actions extends Hyperapp.Actions {
+  reset(): Partial<Module2State>;
+  add(value: number): Partial<Module2State>;
+}
+
+const module2Actions: Hyperapp.ActionsImpl<Module2State, Module2Actions> = {
+  reset: (state, actions) => ({ count: 0 }),
+  add: (state, actions, value: number) => ({ count: state.count + value })
+};
+
+// app
 
 interface State extends Hyperapp.State {
-  count: number
+  module1: Module1State;
+  module2: Module2State;
+  // just to demonstrate there can be other attributes than the ones defined in actions
+  unused?: number;
 }
 
-interface Actions extends Hyperapp.Actions<State> {
-  sub(state: State, actions: ActionCallers, value: number): Partial<State>
-  add(state: State, actions: ActionCallers, value: number): Partial<State>
+interface Actions extends Hyperapp.Actions {
```

Comment:
This looks weird to me. I feel like this should be a generic of state since actions run over state.

Reply:
Yeah, it is, I am trying to do `Action<State>` but I get a lot of issues, and I think it comes from the following truth: State may have any key, which has in turn (almost) any type (VERY easy to write in function of Action). Action has 2 different types of keys: if `key in keyof State`, then it's either a sub-tree of actions, or a library of actions, otherwise it can be any key, in which case it must be of type `(data: any) => ActionResult<...>` (MUCH harder to write in function of State, AND, cannot be an interface, has to be a type...).

Here is the snippet (I think it will be easier to show than try to explain):

```typescript
// represents the external actions exposed to the world
// represents an object where each key (any key) is an action
export type ActionsFunctions<State extends Hyperapp.State> = {
  [action: string]: (data: any) => ActionResult<State>
}

// represents the external actions exposed to the world
// represents an object where each key (MUST be a key in the state object) can be either a sub-tree or a set of actions
export type ActionsTree<State extends Hyperapp.State> = {
  [S in keyof State]?: ActionsTree<State[S]> | ActionsFunctions<State[S]>
}

// represents the external actions exposed to the world
// represents an object where each key is either: an action, a sub-tree or a set of actions
export type Actions<State extends Hyperapp.State> = ActionsFunctions<State> &
  ActionsTree<State>

// represents the internal actions: their implementations
// represents an object where each key (any key) is an action
export type InternalActionsFunctions<
  State extends Hyperapp.State,
  Actions extends Hyperapp.ActionsFunctions<State>
> = {
  [A in keyof Actions]: (
    state: State,
    actions: Actions,
    data: any
  ) => ActionResult<State>
}

// represents the internal actions: their implementations
// represents an object where each key (MUST be a key in the state object) can be either a sub-tree or a set of actions
export type InternalActionsTree<
  State extends Hyperapp.State,
  Actions extends Hyperapp.ActionsTree<State>
> = {
  [S in keyof State]?:
    | InternalActionsTree<State[S], Actions[S]>
    | InternalActionsFunctions<State[S], Actions[S]>
}

// represents the internal actions: their implementations
// represents an object where each key is either: an action, a sub-tree or a set of actions
export type InternalActions<
  State extends Hyperapp.State,
  Actions extends Hyperapp.Actions<State>
> =
  | InternalActionsFunctions<State, Actions>
  | InternalActionsTree<State, Actions>
```

This, however does not work when actually using it:
![screenshot from 2017-09-26 19-59-04](https://user-images.githubusercontent.com/8634093/30875912-3b66e920-a2f5-11e7-9473-2357193d7686.png)

PS: in the screenshot above, even without the error, it gives another error because Module1Actions is an interface and Hyperapp.Actions is a type..

With the approach in this PR, everything just works!

If anyone has any suggestion, feel free to help me out :(

Reply:
Yeah, I think you would want to write that as:

```ts
const module1Actions: Hyperapp.Actions<Module1State> = {
  add: (value: number) => value + 1,
  sub: (value: number) => value - 1,
}
type Module1Actions = typeof module1Actions;
```

Reply:
Check out the new commit, I modified the InternalActions/ActionCallers:
```typescript
export type InternalActions<
  State extends Hyperapp.State & Partial<Record<keyof Actions, any>>,
  Actions extends Hyperapp.Actions<State>
> = {
  [P in keyof Actions]:
    | InternalAction<State, Actions>
    | InternalActions<State[P], Actions[P] & Hyperapp.Actions<State[P]>>
}
```

This `& Partial<Record<keyof Actions, any>>` make it possible to do `InternalActions<State[P], Actions[P] & Hyperapp.Actions<State[P]>>`.

This seems to allow to define actions simply (see index.tsx).